File: python/packages/spore_tools/src/spore_tools/tools/control.py
"""Tier-3 control tools (#81): tools that drive the harness via the escalation
/ clarification protocols rather than returning ordinary data.

Mirrors ``rust/crates/spore-core/src/tools/control.rs``.

* :class:`EnterPlanModeTool` (``enter_plan_mode``) →
  :class:`ToolOutputEscalate` ``{ HarnessSignalEnterPlanMode { context } }``.
* :class:`ExitPlanModeTool` (``exit_plan_mode``) →
  :class:`ToolOutputEscalate` ``{ HarnessSignalExitPlanMode { plan } }``. The
  plan is a structured tool param deserialized DIRECTLY into the existing
  :class:`~spore_core.hooks.PlanArtifact` (issue #81, Q4a — no stub).
* :class:`AskUserQuestionTool` (``ask_user_question``) →
  :class:`ToolOutputAwaitingClarification` ``{ question, options }`` (issue #81,
  Q4b). The harness loop pauses with :class:`HumanRequestClarification`.
* :class:`AbortTool` (``abort``) →
  :class:`ToolOutputEscalate` ``{ HarnessSignalAbort { reason } }``.
"""

from __future__ import annotations

from spore_core.harness import (
    HarnessSignalAbort,
    HarnessSignalEnterPlanMode,
    HarnessSignalExitPlanMode,
    SandboxProvider,
    ToolOutput,
    ToolOutputAwaitingClarification,
    ToolOutputEscalate,
)
from spore_core.model import ToolCall
from spore_core.tool_registry import ToolAnnotations, ToolContext, ToolSchema

from .error import ToolExecutionError
from .params import (
    AbortParams,
    AskUserQuestionParams,
    EnterPlanModeParams,
    ExitPlanModeParams,
    parse_params,
)


# ============================================================================
# EnterPlanMode
# ============================================================================


class EnterPlanModeTool:
    NAME = "enter_plan_mode"

    def name(self) -> str:
        return self.NAME

    def is_subagent_tool(self) -> bool:
        return False

    def may_produce_large_output(self) -> bool:
        return False

    @classmethod
    def schema(cls) -> ToolSchema:
        return ToolSchema(
            name=cls.NAME,
            description="Request entry into plan mode, seeding the planner with context",
            parameters={
                "type": "object",
                "properties": {"context": {"type": "string"}},
            },
            annotations=ToolAnnotations(),
        )

    async def execute(
        self, call: ToolCall, sandbox: SandboxProvider, ctx: ToolContext
    ) -> ToolOutput:
        try:
            params = parse_params(EnterPlanModeParams, call)
        except ToolExecutionError as e:
            return e.to_tool_output()
        return ToolOutputEscalate(signal=HarnessSignalEnterPlanMode(context=params.context))


# ============================================================================
# ExitPlanMode
# ============================================================================


class ExitPlanModeTool:
    NAME = "exit_plan_mode"

    def name(self) -> str:
        return self.NAME

    def is_subagent_tool(self) -> bool:
        return False

    def may_produce_large_output(self) -> bool:
        return False

    @classmethod
    def schema(cls) -> ToolSchema:
        return ToolSchema(
            name=cls.NAME,
            description="Submit the produced plan and request exit from plan mode",
            parameters={
                "type": "object",
                "properties": {
                    "plan": {
                        "type": "object",
                        "properties": {
                            "tasks": {"type": "array", "items": {"type": "string"}},
                            "rationale": {"type": "string"},
                        },
                        "required": ["tasks"],
                    },
                },
                "required": ["plan"],
            },
            annotations=ToolAnnotations(),
        )

    async def execute(
        self, call: ToolCall, sandbox: SandboxProvider, ctx: ToolContext
    ) -> ToolOutput:
        try:
            params = parse_params(ExitPlanModeParams, call)
        except ToolExecutionError as e:
            return e.to_tool_output()
        return ToolOutputEscalate(signal=HarnessSignalExitPlanMode(plan=params.plan))


# ============================================================================
# AskUserQuestion
# ============================================================================


class AskUserQuestionTool:
    NAME = "ask_user_question"

    def name(self) -> str:
        return self.NAME

    def is_subagent_tool(self) -> bool:
        return False

    def may_produce_large_output(self) -> bool:
        return False

    @classmethod
    def schema(cls) -> ToolSchema:
        return ToolSchema(
            name=cls.NAME,
            description="Ask the user a clarifying question (optionally with fixed choices)",
            parameters={
                "type": "object",
                "properties": {
                    "question": {"type": "string"},
                    "options": {"type": "array", "items": {"type": "string"}},
                },
                "required": ["question"],
            },
            annotations=ToolAnnotations(),
        )

    async def execute(
        self, call: ToolCall, sandbox: SandboxProvider, ctx: ToolContext
    ) -> ToolOutput:
        try:
            params = parse_params(AskUserQuestionParams, call)
        except ToolExecutionError as e:
            return e.to_tool_output()
        return ToolOutputAwaitingClarification(question=params.question, options=params.options)


# ============================================================================
# Abort
# ============================================================================


class AbortTool:
    NAME = "abort"

    def name(self) -> str:
        return self.NAME

    def is_subagent_tool(self) -> bool:
        return False

    def may_produce_large_output(self) -> bool:
        return False

    @classmethod
    def schema(cls) -> ToolSchema:
        return ToolSchema(
            name=cls.NAME,
            description="Request a graceful abort of the run with a reason",
            parameters={
                "type": "object",
                "properties": {"reason": {"type": "string"}},
                "required": ["reason"],
            },
            annotations=ToolAnnotations(),
        )

    async def execute(
        self, call: ToolCall, sandbox: SandboxProvider, ctx: ToolContext
    ) -> ToolOutput:
        try:
            params = parse_params(AbortParams, call)
        except ToolExecutionError as e:
            return e.to_tool_output()
        return ToolOutputEscalate(signal=HarnessSignalAbort(reason=params.reason))


__all__ = [
    "AbortTool",
    "AskUserQuestionTool",
    "EnterPlanModeTool",
    "ExitPlanModeTool",
]

File: go/spore-core/tools/control.go
// Tier-3 control tools (#81): tools that drive the harness via the escalation /
// clarification protocols rather than returning ordinary data.
//
//   - EnterPlanModeTool (enter_plan_mode) →
//     ToolOutput.Escalate{ HarnessSignal.EnterPlanMode{ Context } }.
//   - ExitPlanModeTool (exit_plan_mode) →
//     ToolOutput.Escalate{ HarnessSignal.ExitPlanMode{ Plan } }. The plan is a
//     structured tool param deserialized DIRECTLY into the existing
//     sporecore.PlanArtifact (issue #81, Q4a — no stub).
//   - AskUserQuestionTool (ask_user_question) →
//     ToolOutput.AwaitingClarification{ Question, Options } (issue #81, Q4b).
//     The harness loop pauses with HumanRequest.Clarification.
//   - AbortTool (abort) →
//     ToolOutput.Escalate{ HarnessSignal.Abort{ Reason } }.

package tools

import (
	"context"
	"encoding/json"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

// hasField reports whether the JSON object in input contains the named key.
// Used to enforce required fields that Rust gets from serde but Go's
// json.Unmarshal silently defaults — keeping the recoverable-error parity.
func hasField(input []byte, name string) bool {
	var probe map[string]json.RawMessage
	if json.Unmarshal(input, &probe) != nil {
		return false
	}
	_, ok := probe[name]
	return ok
}

// ============================================================================
// EnterPlanMode
// ============================================================================

// EnterPlanModeToolName is the registered tool name.
const EnterPlanModeToolName = "enter_plan_mode"

// EnterPlanModeTool escalates a request to enter plan mode.
type EnterPlanModeTool struct{}

// NewEnterPlanModeTool constructs an EnterPlanModeTool.
func NewEnterPlanModeTool() *EnterPlanModeTool { return &EnterPlanModeTool{} }

func (*EnterPlanModeTool) Name() string                { return EnterPlanModeToolName }
func (*EnterPlanModeTool) IsSubagentTool() bool        { return false }
func (*EnterPlanModeTool) MayProduceLargeOutput() bool { return false }

func (*EnterPlanModeTool) Schema() sporecore.RegistryToolSchema {
	return sporecore.RegistryToolSchema{
		Name:        EnterPlanModeToolName,
		Description: "Request entry into plan mode, seeding the planner with context",
		Parameters: json.RawMessage(`{
			"type": "object",
			"properties": {"context": {"type": "string"}}
		}`),
		Annotations: sporecore.ToolAnnotations{},
	}
}

func (t *EnterPlanModeTool) Execute(_ context.Context, call sporecore.ToolCall, _ sporecore.SandboxProvider, _ *sporecore.ToolContext) sporecore.ToolOutput {
	var params EnterPlanModeParams
	if e := parseParams(call, &params); e != nil {
		return e.ToToolOutput()
	}
	sig := sporecore.HarnessSignal{Kind: sporecore.SignalEnterPlanMode, Context: params.Context}
	return sporecore.ToolOutput{Kind: sporecore.ToolOutputEscalate, Signal: &sig}
}

// ============================================================================
// ExitPlanMode
// ============================================================================

// ExitPlanModeToolName is the registered tool name.
const ExitPlanModeToolName = "exit_plan_mode"

// ExitPlanModeTool escalates the produced plan and requests exit from plan mode.
type ExitPlanModeTool struct{}

// NewExitPlanModeTool constructs an ExitPlanModeTool.
func NewExitPlanModeTool() *ExitPlanModeTool { return &ExitPlanModeTool{} }

func (*ExitPlanModeTool) Name() string                { return ExitPlanModeToolName }
func (*ExitPlanModeTool) IsSubagentTool() bool        { return false }
func (*ExitPlanModeTool) MayProduceLargeOutput() bool { return false }

func (*ExitPlanModeTool) Schema() sporecore.RegistryToolSchema {
	return sporecore.RegistryToolSchema{
		Name:        ExitPlanModeToolName,
		Description: "Submit the produced plan and request exit from plan mode",
		Parameters: json.RawMessage(`{
			"type": "object",
			"properties": {
				"plan": {
					"type": "object",
					"properties": {
						"tasks": {"type": "array", "items": {"type": "string"}},
						"rationale": {"type": "string"}
					},
					"required": ["tasks"]
				}
			},
			"required": ["plan"]
		}`),
		Annotations: sporecore.ToolAnnotations{},
	}
}

func (t *ExitPlanModeTool) Execute(_ context.Context, call sporecore.ToolCall, _ sporecore.SandboxProvider, _ *sporecore.ToolContext) sporecore.ToolOutput {
	var params ExitPlanModeParams
	if e := parseParams(call, &params); e != nil {
		return e.ToToolOutput()
	}
	plan := params.Plan
	sig := sporecore.HarnessSignal{Kind: sporecore.SignalExitPlanMode, Plan: &plan}
	return sporecore.ToolOutput{Kind: sporecore.ToolOutputEscalate, Signal: &sig}
}

// ============================================================================
// AskUserQuestion
// ============================================================================

// AskUserQuestionToolName is the registered tool name.
const AskUserQuestionToolName = "ask_user_question"

// AskUserQuestionTool pauses the harness awaiting a human clarification.
type AskUserQuestionTool struct{}

// NewAskUserQuestionTool constructs an AskUserQuestionTool.
func NewAskUserQuestionTool() *AskUserQuestionTool { return &AskUserQuestionTool{} }

func (*AskUserQuestionTool) Name() string                { return AskUserQuestionToolName }
func (*AskUserQuestionTool) IsSubagentTool() bool        { return false }
func (*AskUserQuestionTool) MayProduceLargeOutput() bool { return false }

func (*AskUserQuestionTool) Schema() sporecore.RegistryToolSchema {
	return sporecore.RegistryToolSchema{
		Name:        AskUserQuestionToolName,
		Description: "Ask the user a clarifying question (optionally with fixed choices)",
		Parameters: json.RawMessage(`{
			"type": "object",
			"properties": {
				"question": {"type": "string"},
				"options": {"type": "array", "items": {"type": "string"}}
			},
			"required": ["question"]
		}`),
		Annotations: sporecore.ToolAnnotations{},
	}
}

func (t *AskUserQuestionTool) Execute(_ context.Context, call sporecore.ToolCall, _ sporecore.SandboxProvider, _ *sporecore.ToolContext) sporecore.ToolOutput {
	var params AskUserQuestionParams
	if e := parseParams(call, &params); e != nil {
		return e.ToToolOutput()
	}
	return sporecore.ToolOutput{
		Kind:     sporecore.ToolOutputAwaitingClarification,
		Question: params.Question,
		Options:  params.Options,
	}
}

// ============================================================================
// Abort
// ============================================================================

// AbortToolName is the registered tool name.
const AbortToolName = "abort"

// AbortTool escalates a graceful-abort signal with a reason.
type AbortTool struct{}

// NewAbortTool constructs an AbortTool.
func NewAbortTool() *AbortTool { return &AbortTool{} }

func (*AbortTool) Name() string                { return AbortToolName }
func (*AbortTool) IsSubagentTool() bool        { return false }
func (*AbortTool) MayProduceLargeOutput() bool { return false }

func (*AbortTool) Schema() sporecore.RegistryToolSchema {
	return sporecore.RegistryToolSchema{
		Name:        AbortToolName,
		Description: "Request a graceful abort of the run with a reason",
		Parameters: json.RawMessage(`{
			"type": "object",
			"properties": {"reason": {"type": "string"}},
			"required": ["reason"]
		}`),
		Annotations: sporecore.ToolAnnotations{},
	}
}

func (t *AbortTool) Execute(_ context.Context, call sporecore.ToolCall, _ sporecore.SandboxProvider, _ *sporecore.ToolContext) sporecore.ToolOutput {
	var params AbortParams
	if e := parseParams(call, &params); e != nil {
		return e.ToToolOutput()
	}
	// reason is required (Rust serde rejects its absence; Go json.Unmarshal does
	// not, so validate explicitly to keep the recoverable-error parity).
	if !hasField(call.Input, "reason") {
		return InvalidParameters("missing required field `reason`").ToToolOutput()
	}
	sig := sporecore.HarnessSignal{Kind: sporecore.SignalAbort, Reason: params.Reason}
	return sporecore.ToolOutput{Kind: sporecore.ToolOutputEscalate, Signal: &sig}
}

var (
	_ sporecore.Tool = (*EnterPlanModeTool)(nil)
	_ sporecore.Tool = (*ExitPlanModeTool)(nil)
	_ sporecore.Tool = (*AskUserQuestionTool)(nil)
	_ sporecore.Tool = (*AbortTool)(nil)
)
